neuraxle/base.py

```diff
@@ -135,70 +140,255 @@ def concat(self, data_container: DataContainer):
         self.expected_outputs.extend(data_container.expected_outputs)
 
 
+class BaseSaver(ABC):
+    """
+    Any saver must inherit from this one. Some savers just save parts of objects, some save it all or what remains.
+    """
+
+    @abstractmethod
+    def save_step(self, step: 'BaseStep', context: 'ExecutionContext') -> 'BaseStep':
+        """
+        Save step with execution context.
+
+        :param step: step to save
+        :param context: execution context
+        :return:
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def can_load(self, step: 'BaseStep', context: 'ExecutionContext'):
+        raise NotImplementedError()
+
+    @abstractmethod
+    def load_step(self, step: 'BaseStep', context: 'ExecutionContext') -> 'BaseStep':
+        """
+        Load step with execution context.
+
+        :param step: step to load
+        :param context: execution context to load from
+        :return: loaded base step
+        """
+        raise NotImplementedError()
+
+
+class JoblibStepSaver(BaseSaver):
+    """
+    This saver is a good default saver when the object is
+    already stripped out of things that would make it unserializable.
+    """
+
+    def can_load(self, step: 'BaseStep', context: 'ExecutionContext'):
+        return os.path.exists(
+            os.path.join(context.get_path(), '{0}.joblib'.format(step.name))
+        )
+
+    def save_step(self, step: 'BaseStep', context: 'ExecutionContext') -> 'BaseStep':
+        """
+        Saved step stripped out of things that would make it unserializable.
+
+        :param step: stripped step to save
+        :param context: execution context to save from
+        :return:
+        """
+        if not os.path.exists(context.get_path()):
+            context.mkdir()
+
+        path = os.path.join(context.get_path(), '{0}.joblib'.format(step.name))
+        dump(step, path)
+
+        return step
+
+    def load_step(self, step: 'BaseStep', context: 'ExecutionContext') -> 'BaseStep':
+        """
+        Load stripped step.
+
+        :param step: stripped step to load
+        :param context: execution context to load from
+        :return:
+        """
+        loaded_step = load(os.path.join(context.get_path(), '{0}.joblib'.format(step.name)))
+
+        # we need to keep the current steps in memory because they have been deleted before saving...
+        # the steps that have not been saved yet need to be in memory while loading a truncable steps...
+        if isinstance(step, TruncableSteps):
+            loaded_step.steps = step.steps
```

Comment:
this is really dirty, any suggestions to make this cleaner ? @guillaume-chevalier